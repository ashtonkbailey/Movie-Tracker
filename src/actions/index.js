// movie actions
/////// add movies
export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
})

// user actions
//////// add favorite, delete favorite, sign in, sign out, create user

// export const addUser = (userObject) => ({
//   type: 'ADD_USER',
//   user
// })

// export const logInUser = (userObject) => ({
//   type: 'LOG_IN_USER',
//   user
// })