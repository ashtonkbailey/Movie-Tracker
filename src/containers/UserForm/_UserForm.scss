.user-form {
  background: #750413;
  width: 60%;
  max-width: 900px;
  min-width: 350px;
  height: 60vh;
  margin: auto;
}

.header-text {
  margin: auto;
  height: 60px;
  color: #FEEB9E;
  font-family: 'Sanchez', serif;
}

form {
  display: flex;
  margin: auto;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  height: 290px;
  width: 250px;

  input {
    width: 100%;
    height: 27px;
    font-size: 1rem;
    font-family: 'Sanchez', serif;
    padding-left: 6px;
    outline: none;

    &::placeholder {
      font-size: 1rem;
    }
  }

  button {
    background: #FEEB9E;
    border-radius: 1px;
    border: none;
    outline: none;
    height: 27px;
    width: 70%;
    font-size: 1rem;
    font-family: 'Sanchez', serif;
    transition: 0.4s ease;
    cursor: pointer;

    &:hover {
      font-size: 1.05rem;
      text-shadow: 0 0 3px rgba(0, 0, 0, 0.4);
      box-shadow:
        0 0 5px rgba(254, 235, 158, 1),
        0 0 0px rgba(254, 235, 158, 1),
        0 0 10px rgba(254, 235, 158, 1);
    }
  }
}

.user-msg {
  color: #FEEB9E;
  font-family: 'Sanchez', serif;
  text-align: center;
  opacity: 0;
  transition: 0.2s;
}

.error {
  opacity: 1;
}

.no-acct-msg {
  width: 300px;
  text-align: center;
  color: #FEEB9E;
  font-family: 'Sanchez', serif;
  font-size: 1rem;
  cursor: pointer;
}

span.signup-here {
  text-decoration: underline;
}