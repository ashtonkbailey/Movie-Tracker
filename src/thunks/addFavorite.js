export const addFavoriteThunk = (favoriteObj, favorites) => {
  return async (dispatch) => {
    try {
      if (favorites.length) {
        let alreadyAFav = favorites.find(favorite => {
          return favorite === favoriteObj.movie_id
        })
        if (alreadyAFav) {
          throw new Error('Already a favorite')
        }
      }
      const url = 'http://localhost:3000/api/users/favorites/new'
      const result = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(favoriteObj),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if (!result.ok) {
        throw new Error('Unable to save to favorites')
      }
    } catch (error) {
      console.log(error)
    }
  }
}
