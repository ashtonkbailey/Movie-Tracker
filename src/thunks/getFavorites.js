import { getFavorites } from '../actions/index'

export const getFavoritesThunk = (userId) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/users/${userId}/favorites`
      const result = await fetch(url)
      if (!result.ok) {
        throw new Error('Unable to get your favorites')
      }
      const favorites = await result.json()
      const favoriteIds = favorites.data.map(favorite => favorite.movie_id)
      dispatch(getFavorites(favoriteIds))
    } catch (error) {
      console.log(error)
    }
  }
}
