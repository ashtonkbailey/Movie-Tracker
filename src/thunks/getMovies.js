import { key } from '../utils/apiKey';
import {setError} from '../actions'

export const getMoviesThunk = () => {
  return async (dispatch) => {
    try {
      const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${key}&language=en-US&page=1`;
      const results = await fetch(url);
      if(!results.ok) {
        throw Error('Could not get movies')
      }
      const movieData = await results.json();
      return movieData.results
    } catch(error) {
      dispatch(setError(error.message))

    }
  }
}
