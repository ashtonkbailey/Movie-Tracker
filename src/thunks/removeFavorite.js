import { setError } from '../actions'

export const removeFavoriteThunk = (id, userId) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/users/${userId}/favorites/${id}`;
      const response = await fetch(url, {
        method: 'DELETE',
        body: JSON.stringify({movie_id: id, id: userId}),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if (!response.ok) {
        throw Error ('Unable to remove favorite')
      }
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}