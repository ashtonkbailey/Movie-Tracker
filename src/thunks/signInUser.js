import { logInUser, setError } from '../actions/index'

export const signInUserThunk = (user) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/users`
      const result = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          'Content-type': 'application/json'
        }
      })
      if (!result.ok) {
        throw new Error('Email or password do not match!')
      }
      const data = await result.json()
      dispatch(logInUser(data.data))
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}