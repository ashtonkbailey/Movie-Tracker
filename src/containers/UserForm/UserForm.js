import React, { Component } from 'react';
import { connect } from 'react-redux';
import '../../index.scss';
import * as actions from '../../actions/index';
import { Redirect, withRouter, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { getFavoritesThunk } from '../../thunks/getFavorites';
import { addNewUserThunk } from '../../thunks/addNewUser';
import { signInUserThunk } from '../../thunks/signInUser';

export class UserForm extends Component {
  constructor() {
    super()
    this.state = {
      name: '',
      email: '',
      password: '',
      loggedIn: false,
      signInError: false,
      logInError: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({[name]: value})
  }

  handleSubmit = async (e, type) => {
    e.preventDefault()
    const newUser = {...this.state, loggedIn: true}
    type === 'login'
      ? await this.handleLogin(newUser)
      : await this.handleNewUser(newUser)
    if (!this.state.signInError && !this.state.logInError){
      this.props.getFavoritesThunk(this.props.user.id)
    }
  }

  handleLogin = async (newUser) => {
    newUser.email = newUser.email.toLowerCase()
    try {
      
      const loggedInUser = await this.props.signInUser(newUser)

      this.setState({ loggedIn: true })
    } catch (error) {
      this.setState({ password: '', logInError: true })
    }
  }

  handleNewUser = async (newUser) => {
    try {
      const userId = await this.props.addNewUser(newUser)
      this.setState({ loggedIn: true })
    } catch (error) {
      this.setState({ email: '', password: '', signInError: true })
    }
  }

  render() {
    const { type } = this.props
    const { signInError, logInError } = this.state
    if (this.state.loggedIn) {
      return <Redirect to='/' />
    }

    let errorText
    if (signInError) {
      errorText = 'User email already exists! Please log in or try a different email.'
    } else if (logInError) {
      errorText = 'Email and password do not match.'
    }

    let error = false
    if (signInError || logInError) {
      error = true
    }

    let button
    type === 'login'
      ? button = 'Log In'
      : button = 'Sign Up'

    let headerText
    type === 'login'
      ? headerText = 'Log In'
      : headerText = 'Sign Up'

    let nameInput
    type === 'login'
    ? nameInput = (
      <Link to='/signup'>
        <div className="no-acct-msg">Don't have an account? Sign up <span className="signup-here">here</span>.
        </div>
      </Link>
    )
    : nameInput = (
        <input onChange={this.handleChange}
          name='name'
          value={this.state.name}
          placeholder='Name (optional)'
          type='text'
        />
      )

    return (
      <div className='user-form'>
        <h1 className="header-text">{headerText}</h1>
        <form onSubmit={(e) => this.handleSubmit(e, type)}>
          {nameInput}
          <input onChange={this.handleChange}
            name='email'
            value={this.state.email}
            placeholder='E-mail'
            type='email'
            autoFocus={true}
          />
          <input onChange={this.handleChange}
            name='password'
            value={this.state.password}
            placeholder='Password'
            type='password'
          />
          <button>{button}</button>
        </form>
        <p className={`user-msg ${error && 'error'}`}>{errorText}</p>
      </div>
    )
  }
}

UserForm.propTypes = {
  signInUser: PropTypes.func.isRequired,
  addNewUser: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  getFavoritesThunk: PropTypes.func.isRequired
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  signInUser: (user) => dispatch(signInUserThunk(user)),
  addNewUser: (user) => dispatch(addNewUserThunk(user)),
  getFavoritesThunk: (userId) => dispatch(getFavoritesThunk(userId))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(UserForm));
