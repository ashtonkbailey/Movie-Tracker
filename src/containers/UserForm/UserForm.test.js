import { UserForm, mapDispatchToProps } from './UserForm';
import * as actions from '../../actions/index';
import { shallow } from 'enzyme';
import React from 'react';
import { signInUserThunk } from '../../thunks/signInUser'
import { addNewUserThunk } from '../../thunks/addNewUser'

jest.mock('../../thunks/signInUser')
jest.mock('../../thunks/addNewUser')


describe('UserForm', () => {
  describe('mapDispatchToProps', () => {
    it('should return a props object with the method addNewUser', () => {
      const mockDispatch = jest.fn();
      const mockUser = {name: 'bob', email: 'bob@example.com', password: 'password'}
      const action = addNewUserThunk(mockUser)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.addNewUser(mockUser)
      expect(mockDispatch).toHaveBeenCalledWith(action)
    })
  })

  describe('UserForm Component', () => {
    it('should match the snapshot', () => {
      const mockAddNewUser = jest.fn()
      const mockSignInUser = jest.fn()
      const mockUser = {name: 'Sam'}
      const mockGetFavoritesThunk = jest.fn()

      const wrapper = shallow(<UserForm  user={mockUser} getFavoritesThunk={mockGetFavoritesThunk} addNewUser={mockAddNewUser} signInUser={mockSignInUser}/>)
      expect(wrapper).toMatchSnapshot()
    })

    it('should have the correct default state', () => {
      const mockUser = {}
      const mockGetFavoritesThunk = jest.fn()
      const expected = {
      name: '',
      email: '',
      password: '',
      loggedIn: false,
      signInError: false,
      logInError: false
      }
      const mockAddNewUser = jest.fn()
      const mockSignInUser = jest.fn()
      const wrapper = shallow(<UserForm  user={mockUser} addNewUser={mockAddNewUser}signInUser={mockSignInUser} getFavoritesThunk={mockGetFavoritesThunk} />)
      expect(wrapper.state()).toEqual(expected)
    })

    it('should change the value of state when handleChange is called', () => {
      const mockAddNewUser = jest.fn()
      const mockSignInUser = jest.fn()
      const mockEvent = {target: {name: 'userName', value: 'sally'}}
      const mockUser = {}
      const mockGetFavoritesThunk = jest.fn()
      const wrapper = shallow(<UserForm  addNewUser={mockAddNewUser}signInUser={mockSignInUser} user={mockUser} getFavoritesThunk={mockGetFavoritesThunk} />)
      const expected = 'sally'
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state().userName).toEqual(expected)
    })

    describe('handleSubmit', () => {
      it('should call handleLogin', async() => {
        const mockAddNewUser = jest.fn()
        const mockSignInUser = jest.fn()
        const mockUser = {}
        const mockGetFavoritesThunk = jest.fn()
        const mockEvent = { preventDefault: jest.fn() };
        const wrapper = shallow(<UserForm addNewUser={mockAddNewUser}signInUser={mockSignInUser}
          type="login"  user={mockUser} getFavoritesThunk={mockGetFavoritesThunk}
        />);
        wrapper.instance().handleLogin = jest.fn();
        await wrapper.instance().handleSubmit(mockEvent, 'login');
        expect(wrapper.instance().handleLogin).toHaveBeenCalled();
      });

      it('should call HandleNewUser', async() => {
        const mockAddNewUser = jest.fn()
        const mockSignInUser = jest.fn()
        const mockUser = {}
        const mockGetFavoritesThunk = jest.fn()
        const mockEvent = { preventDefault: jest.fn() };
        const wrapper = shallow(<UserForm
          type="signup" addNewUser={mockAddNewUser}signInUser={mockSignInUser}user={mockUser} getFavoritesThunk={mockGetFavoritesThunk}
        />);
        wrapper.instance().handleNewUser = jest.fn();
        await wrapper.instance().handleSubmit(mockEvent, 'signup');
        expect(wrapper.instance().handleNewUser).toHaveBeenCalled();
      });


      it('should run handleSubmit on form submit', () => {        
        const mockAddNewUser = jest.fn()
        const mockSignInUser = jest.fn()
        const mockUser = {}
        const mockGetFavoritesThunk = jest.fn()
        const wrapper = shallow(<UserForm
          type="login" addNewUser={mockAddNewUser}signInUser={mockSignInUser} user={mockUser} getFavoritesThunk={mockGetFavoritesThunk}
        />);
        const e = { preventDefault: jest.fn() }
        wrapper.instance().handleSubmit = jest.fn()

        wrapper.find('form').simulate('submit', e)

        expect(wrapper.instance().handleSubmit).toBeCalledWith(e, 'login')
      })
    })
  })

  describe('handleLogin', () => {
    it('should update state for logged in user', async () => {
      const mockNewUser = {name: 'bob', password: '123', email: 'bob@goodlife.com'}
      signInUserThunk.mockImplementation(()=> {
        return mockNewUser
      })
      const mockAddNewUser = jest.fn()
      const mockSignInUser = jest.fn()
      const mockUser = {}
      const mockGetFavoritesThunk = jest.fn()
      const wrapper = shallow(<UserForm
        addNewUser={mockAddNewUser}signInUser={mockSignInUser} user={mockUser} getFavoritesThunk={mockGetFavoritesThunk}
        // type="login"
      />);
      const expected = true;
      await wrapper.instance().handleLogin(mockNewUser);
      expect(wrapper.state().loggedIn).toEqual(expected);
})
  })

})