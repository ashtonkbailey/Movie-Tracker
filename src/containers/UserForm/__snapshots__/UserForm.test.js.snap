// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm UserForm Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="header-text"
        >
          Sign Up
        </h1>,
        <form
          onSubmit={[Function]}
        >
          <input
            name="name"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
          <input
            name="email"
            onChange={[Function]}
            placeholder="E-mail"
            type="email"
            value=""
          />
          <input
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <button>
            Sign Up
          </button>
        </form>,
        <p
          className="user-msg false"
        />,
      ],
      "className": "user-form",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign Up",
          "className": "header-text",
        },
        "ref": null,
        "rendered": "Sign Up",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />,
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />,
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />,
            <button>
              Sign Up
            </button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "E-mail",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up",
            },
            "ref": null,
            "rendered": "Sign Up",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "user-msg false",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="header-text"
          >
            Sign Up
          </h1>,
          <form
            onSubmit={[Function]}
          >
            <input
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
            <button>
              Sign Up
            </button>
          </form>,
          <p
            className="user-msg false"
          />,
        ],
        "className": "user-form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign Up",
            "className": "header-text",
          },
          "ref": null,
          "rendered": "Sign Up",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
              />,
              <input
                name="email"
                onChange={[Function]}
                placeholder="E-mail"
                type="email"
                value=""
              />,
              <input
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />,
              <button>
                Sign Up
              </button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "name",
                "onChange": [Function],
                "placeholder": "Name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "E-mail",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign Up",
              },
              "ref": null,
              "rendered": "Sign Up",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "user-msg false",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
